### ISSUE TITLE
Nested outputs structures and corresponding nested losses are supported, but not nested metrics

### ISSUE THREAD

**Issue by hertschuh** [keras-team/keras#21700]
This is a follow up to https://github.com/keras-team/keras/issues/21675 and https://github.com/keras-team/keras/pull/21694

- nested output structures are supported
- nested losses matching the nested output structures are supported
- however, the same is not true for metrics, the code within the `try`/`except` block fails

```python
import numpy as np
import keras

print('keras version:', keras.__version__)

X = np.random.rand(100, 32)
Y1 = np.random.rand(100, 10)
Y2 = np.random.rand(100, 10)
Y3 = np.random.rand(100, 10)

def create_model():
    x = keras.Input(shape=(32,))
    y1 = keras.layers.Dense(10)(x)
    y2 = keras.layers.Dense(10)(x)
    y3 = keras.layers.Dense(10)(x)
    return keras.Model(inputs = x, outputs = {'a': y1, 'b': {'c': y2, 'd': y3}})

model = create_model()
model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}})
model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
print('Successful training not using metrics\n')

try:
    model = create_model()
    model.compile(
        optimizer = 'adam',
        loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}},
        metrics = {'a': ['mae', 'acc'], 'b':{'c': 'mse', 'd': 'mse'}}
    )
    model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
    print('Successful training using metrics')
except ValueError as ex:
    print('Unsuccessful training with metrics')
    print(ex)

model = create_model()
model.compile(
    optimizer = 'adam',
    loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}},
    loss_weights = {'a': 1., 'b': {'c': 0.5, 'd': 0.4}}
)
model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
print('Successful training using loss_weights\n')
```

Ideally, metrics and losses should be handled the same. Part of the complication is that there is a mechanism to use the output names when not using the output structure.

Another complication is that metrics can be either a single metric or a list of metrics for each output. The logic that turns single metrics to a list must handle nested structures correctly (with `keras.tree.map_structure_up_to`).

**Comment by tripathi-genius** [comment]
Are you still taking contributions?

**Comment by bitloi** [comment]
Hi @hertschuh I’d like to work on the issue and would like to confirm the approach before opening a PR.

**Understanding:** Nested output structures and nested losses are already handled in `CompileLoss` via `_build_nested()` and tree traversal, but `CompileMetrics` only supports a single level of dict keys and treats each `metrics[name]` as a list of metric specs, so nested configs like `{'a': ['mae'], 'b': {'c': 'mse', 'd': 'mse'}}` fail.

**Proposed approach:**

1. **`CompileMetrics.build()`**  
   For nested `y_pred`, resolve output names and flattening in the same way as `CompileLoss`: use `tree.flatten(y_pred)` and align with the same flat order (e.g. via `tree.pack_sequence_as` with the model’s output structure) so we have a well-defined list of (y_true_i, y_pred_i) per output leaf.

2. **`_build_metrics_set()` (or a small private helper)**  
   When `metrics` is a dict, support nested dict/list structure by recursing over the structure (similar to `CompileLoss._build_nested()`), so that at each leaf we have either a single metric or a list of metrics. Use `tree.map_structure_up_to` (or equivalent recursion) to normalize “single metric or list” per leaf, then build one `MetricsList` per leaf and append to `_flat_metrics` in the **same order** as `tree.flatten(y_pred)` so alignment with loss behavior is preserved.

3. **`_flatten_y()`**  
   For nested outputs, return `tree.flatten(y)` so that the list length and order match `_flat_metrics` and `update_state`’s `zip(_flat_metrics, y_true, y_pred)` remains correct.

4. **Output names**  
   Keep compatibility with the existing “output names when not using output structure” behavior, and ensure the logic that turns a single metric into a list respects nested structure (as in the issue description).

**Scope:** Changes only in `keras/src/trainers/compile_utils.py` (no new public API). I’ll add tests mirroring the existing nested-loss cases (e.g. `test_struct_loss`-style) for nested metrics.

Does this direction align with what you have in mind? Any constraints or preferences I should follow before opening a PR?

**Comment by hertschuh** [comment]
@bitloi there are already 2 PRs open about this (I haven't reviewed them).
